orca-api/src/main/java/com/netflix/spinnaker/orca/api/pipeline/remote/execution/RemoteStageExecution.java

```diff
@@ -0,0 +1,41 @@
+package com.netflix.spinnaker.orca.api.pipeline.remote.execution;
+
+import com.netflix.spinnaker.kork.annotations.Alpha;
+import lombok.Builder;
+import lombok.Data;
+
+import javax.annotation.Nonnull;
+import java.util.Map;
+
+/**
+ * This object is sent to the remote stage target when initializing the stage.
+ */
+@Alpha
+@Data
+@Builder
+public class RemoteStageExecution {
+
+  /** The type identifier of the stage. */
+  @Nonnull
+  private String type;
+
+  /** The stage execution ID as generated by Orca. */
+  @Nonnull
+  private String id;
+
+  /** The pipeline execution ID which the stage belongs to. */
+  @Nonnull
+  private String pipelineExecutionId;
+
+  /** The stage context provided at stage initialization. */
+  @Nonnull
+  private Map<String, Object> context;
+
+  /** All upstream stage outputs ordered by stage refId. */
+  @Nonnull
+  private Map<String, Map<String, Object>> upstreamStageOutputs;
```

Comment:
I'm wondering if it would be better to hide the details of upstream outputs (from the stage author) and just emulate what we do in StageContext?

https://github.com/spinnaker/orca/blob/ae2b8043e538b3ed0db39b61c8da0d2e36ee8a5a/orca-core/src/main/java/com/netflix/spinnaker/orca/pipeline/model/StageContext.java#L57-L67

Reply:
Hm, that's an interesting idea.  So basically just provide the `context` which is the merger of the current stage (in `StageContext` represented by `delegate`) and the `ancestors` stage outputs?  I think the key in the current stage context would have to override a matching key in the ancestors outputs.

Reply:
Traverse the ancestors backwards and add all stage context keys to a map: 

```java
      ArrayList<StageExecution> ancestors = new ArrayList<>(stage.ancestors());
      Collections.reverse(ancestors);
      Map<String, Object> result = new HashMap<>();
      ancestors.forEach(stage -> result.putAll(stage.getContext()));
```

... I think

Reply:
Yeah that looks legit-ish to me. I'll try it out - thanks!

Reply:
Curious -- this could be quite large, any restrictions on actually getting it into the target runtime?

In some future time, it'd be neat to allow for a stage to define _what_ data it'd like passed to it as a more strongly defined contract. If that data's not available, than we fail.

Reply:
We could do that instead, and then just provide a mechanism for querying for upstream stage outputs.  Maybe that's better?

Reply:
> provide a mechanism for querying for upstream stage outputs

Meaning, a remote stage would say, "I need context on X, Y, Z" and then the query mechanism would satisfy this and pass the results into the remote stage inputs, or were you thinking something else?

Reply:
> Which bake stage wins?

What would currently happen for in-process stages? 

---

In general, I think we should minimize what we pass to the stage.  If they need to do fancy things like query the execution graph, then maybe an escape hatch would be to fetch that execution from gate (within the remote stage).

Reply:
@robzienert That is basically what I was thinking ... perhaps with some formalization around certain types of data we already know about (deployment details, etc.).

Reply:
one more in favor of passing in concrete type (and not allowing the stage to look stuff up) if we "leak" the "lookup something from some random stage" concept into plugins it will be that much harder for us to refactor this painful aspect of orca later.